### PR TITLE
Rodrigo/testcases

### PR DESCRIPTION
**Pull request by rodrigovalle**
Added testcases mostly, changed some docker integration stuff too :<

Build is failing, not sure why. I can't run the tests locally.

### CODE REVIEW

File: routes/tests.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+/*
+const assignment = await db.Assignment.findById(assignid).exec();
+if (!assignment) {
+    res.status(404);
+    return res.render('error', {message: "Assignment not found."});
+}
+*/
+
+router.get('/', async (req, res) => {
+    if (!isAuthorized(req.query.assignid, res.locals.user)){
```
Comment (abalbanyan) [inline routes/tests.js]: isAuthorized needs to be "await"ed

File: routes/tests.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+var router =  express.Router();
+var db = require('../db.js');
+var { isCourseInstructor } = db.utils;
+var _ = require('lodash/core');
+
+async function isAuthorized(assignId, user) {
+    const userId = user.userid;
+    const isInstructor = user.instructor;
+
+    return assignId && isInstructor && isCourseInstructor(
```
Comment (abalbanyan) [inline routes/tests.js]: "isCourseInstructor" needs to be awaited

File: routes/tests.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+*/
+
+router.get('/', async (req, res) => {
+    if (!isAuthorized(req.query.assignid, res.locals.user)){
+        return res.status(403).render('error', {
+            message: "You do not have access to this assignment."
+        });
+    }
+    
+    let tests = await db.TestCase.find({ assignid: req.query.assignid });
```
Comment (abalbanyan) [inline routes/tests.js]: add ".exec()" to the end of this query.

File: routes/tests.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+router.post('/', async (req, res) => {
+    let required = ['name', 'stdin', 'stdout'];
+    if (!required.every(prop => prop in req.body)) {
+        return res.status(400).json({
+            name: 'Error',
+            message: 'Not all required fields were given'
+        });
+    }
+
+    if (!isAuthorized(req.query.assignid, res.locals.user)) {
```
Comment (abalbanyan) [inline routes/tests.js]: needs to be awaited

File: routes/tests.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        });
+    }
+
+    try {
+        await db.TestCase.create({
+            assignid: req.query.assignid,
+            name: req.body.name,
+            stdin: req.body.stdin,
+            stdout: req.body.stdout
+        });
```
Comment (abalbanyan) [inline routes/tests.js]: add .exec() to the end of this query

File: routes/tests.js [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    let required = ['testid'];
+    if (!required.every(prop => prop in req.body)) {
+        return res.status(400).json({
+            name: 'Error',
+            message: 'Not all required fields were given'
+        });
+    }
+
+    let testcase = await db.TestCase.findById(req.body.testid);
+    if (!isAuthorized(testcase.assignid, res.locals.user)) {
```
Comment (abalbanyan) [inline routes/tests.js]: await isAuthorized

File: routes/tests.js [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+router.put('/', async (req, res) => {
+    let required = ['testid'];
+    if (!required.every(prop => prop in req.body)) {
+        return res.status(400).json({
+            name: 'Error',
+            message: 'Not all required fields were given'
+        });
+    }
+
+    let testcase = await db.TestCase.findById(req.body.testid);
```
Comment (abalbanyan) [inline routes/tests.js]: add .exec() to the end of this query

File: routes/tests.js [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    if (!isAuthorized(testcase.assignid, res.locals.user)) {
```
Comment (abalbanyan) [inline routes/tests.js]: await isAuthorized

File: routes/tests.js [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        return res.status(401).json({
+            name: 'Error',
+            message: 'Unauthorized to delete testcases'
+        });
+    }
+
+    try {
+        await db.TestCase.findByIdAndDelete(testcase.id);
```
Comment (abalbanyan) [inline routes/tests.js]: add .exec() to the end of this query

**Review by abalbanyan — CHANGES_REQUESTED** [review]

File: routes/tests.js (hunk 9 above)
Comment (rodrigovalle) [reply routes/tests.js]: how did any of this work in the first place

**Pull request merged**